include/fiction/algorithms/physical_design/wiring_reduction.hpp

```diff
@@ -0,0 +1,935 @@
+            }
+        };
+
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(cartesian_layout<OffsetCoordinateType>::east(c)));
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(c));
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(cartesian_layout<OffsetCoordinateType>::west(c)));
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::east(c));
+    }
+    /**
+     * Iterates over adjacent coordinates of a given coordinate in the middle rows.
+     *
+     * This function iterates over adjacent coordinates of the given coordinate `c` in the middle rows
+     * and applies the provided functor `fn` to each valid adjacent coordinate.
+     *
+     * @tparam Fn Type of the functor to apply to each adjacent coordinate.
+     * @param c The reference coordinate for which adjacent coordinates are determined.
+     * @param fn The functor to apply to each adjacent coordinate.
+     */
+    template <typename Fn>
+    void foreach_adjacent_coordinate_middle_rows(const OffsetCoordinateType& c, Fn&& fn) const
+    {
+        const auto apply_if_not_c = [&c, &fn](const auto& cardinal) noexcept
+        {
+            if (cardinal != c)
+            {
+                std::invoke(std::forward<Fn>(fn), cardinal);
+            }
+        };
+
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(cartesian_layout<OffsetCoordinateType>::east(c)));
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(c));
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::south(cartesian_layout<OffsetCoordinateType>::west(c)));
+    }
+    /**
+     * Iterates over adjacent coordinates of a given coordinate in the last row.
+     *
+     * This function iterates over adjacent coordinates of the given coordinate `c` in the last row
+     * and applies the provided functor `fn` to each valid adjacent coordinate.
+     *
+     * @tparam Fn Type of the functor to apply to each adjacent coordinate.
+     * @param c The reference coordinate for which adjacent coordinates are determined.
+     * @param fn The functor to apply to each adjacent coordinate.
+     */
+    template <typename Fn>
+    void foreach_adjacent_coordinate_last_row(const OffsetCoordinateType& c, Fn&& fn) const
+    {
+        const auto apply_if_not_c = [&c, &fn](const auto& cardinal) noexcept
+        {
+            if (cardinal != c)
+            {
+                std::invoke(std::forward<Fn>(fn), cardinal);
+            }
+        };
+
+        apply_if_not_c(cartesian_layout<OffsetCoordinateType>::east(c));
+    }
+
+  private:
+    /**
+     * The current search direction: horizontal (from left to right) and vertical (from top to bottom).
+     */
+    search_direction search_dir;
+};
+
+/**
+ * Type alias for an obstruction layout specialized for finding excess wiring.
+ */
+using wiring_reduction_lyt = obstruction_layout<wiring_reduction_layout<>>;
```

Comment:
This change did not address my comment as I intended it. I was wondering if we should use `using wiring_reduction_lyt = obstruction_layout<wiring_reduction_layout<OffsetCoordinateType>>;` here instead. Otherwise, the coordinate types might differ.